### PR TITLE
AWS: initial multi account client setup

### PR DESCRIPTION
**Pull request by mikecutalo**
### Description
In an effort to break up the multi account support to smaller PR's so they are easier to review. In this PR I incrementally move forward for multi account AWS support, below is a summary of whats changed.

* Introduction of account primitive with new structs
* Generate all clients for all accounts
* Update all test to compensate for account
* Introduced some half step code satisfying existing functionality before the final cut over to multi account.

### Testing Performed
Unit / Local.

### GitHub Issue
#1853

### CODE REVIEW

File: backend/service/aws/aws.go [hunk 1]
```diff
@@ -48,12 +50,18 @@ func New(cfg *anypb.Any, logger *zap.Logger, scope tally.Scope) (service.Service
 		return nil, err
 	}
```
Comment (danielhochman) [inline backend/service/aws/aws.go:52]: throw unimplemented if profile config option is set?

File: backend/service/aws/aws.go [hunk 2]
```diff
@@ -80,29 +87,88 @@ func New(cfg *anypb.Any, logger *zap.Logger, scope tally.Scope) (service.Service
...
+	}
+
+	if err := c.configureAdditonalAccountClient(ac.AdditionalAccounts, ds, awsHTTPClient); err != nil {
+		return nil, err
 	}
 
 	return c, nil
 }
 
+func (c *client) configureAdditonalAccountClient(accounts []*awsv1.AWSAccount, ds *awsv1.ScalingLimits, awsHTTPClient *http.Client) error {
```
Comment (danielhochman) [inline backend/service/aws/aws.go:108]: nit: spelling
```suggestion
func (c *client) configureAdditionalAccountClient(accounts []*awsv1.AWSAccount, ds *awsv1.ScalingLimits, awsHTTPClient *http.Client) error {
```

File: backend/service/aws/aws.go [hunk 3]
```diff
@@ -262,10 +341,19 @@ func newProtoForAutoscalingGroup(group astypes.AutoScalingGroup) *ec2v1.Autoscal
...
-	for region := range c.clients {
+	uniqueRegions := map[string]bool{}
+
+	for _, account := range c.accounts {
+		for region := range account.clients {
+			uniqueRegions[region] = true
+		}
+	}
+
+	regions := []string{}
```
Comment (danielhochman) [inline backend/service/aws/aws.go:352]: super nit, you can `make` with len

File: backend/service/aws/aws.go [hunk 4]
```diff
@@ -80,29 +87,88 @@ func New(cfg *anypb.Any, logger *zap.Logger, scope tally.Scope) (service.Service
...
+	for _, account := range accounts {
+		accountRoleARN := fmt.Sprintf("arn:aws:iam::%s:role/%s", account.AccountNumber, account.IamRole)
+		// For doing STS calls it does not matter which region client we are using, as they are not bounded by region
+		// we choose just the first region client
+		stsClient := c.accounts[c.currentAccountAlias].clients[c.accounts[c.currentAccountAlias].regions[0]].sts
+		assumeRoleProvider := stscreds.NewAssumeRoleProvider(stsClient, accountRoleARN)
+		credsCache := aws.NewCredentialsCache(assumeRoleProvider)
+
+		for _, region := range account.Regions {
+			regionCfg, err := config.LoadDefaultConfig(context.TODO(),
```
Comment (danielhochman) [inline backend/service/aws/aws.go:118]: any thoughts on how to reduce the duplication of config creation code on line :76?

**Review by danielhochman — COMMENTED** [review]
great, thanks for splitting into smaller chunks. some initial comments.

File: backend/service/aws/aws.go (hunk 4 above)
Comment (mikecutalo) [reply backend/service/aws/aws.go:119]: Yeah played around with it a bit and this should reduce some of the duplication and make it easier to handle common options, let me know what you think.

File: backend/service/aws/aws.go [hunk 5]
```diff
@@ -62,47 +77,90 @@ func New(cfg *anypb.Any, logger *zap.Logger, scope tally.Scope) (service.Service
 	}
 
 	ds := getScalingLimits(ac)
-
 	awsHTTPClient := &http.Client{}
+	awsClientCommonOptions := []func(*config.LoadOptions) error{
```
Comment (danielhochman) [inline backend/service/aws/aws.go:81]: nit: use `LoadOptionsFunc`

File: backend/service/aws/aws.go [hunk 6]
```diff
@@ -62,47 +77,90 @@ func New(cfg *anypb.Any, logger *zap.Logger, scope tally.Scope) (service.Service
...
-			sts:         sts.NewFromConfig(regionCfg),
-		}
+	if err := c.configureAdditionalAccountClient(ac.AdditionalAccounts, ds, awsHTTPClient, awsClientCommonOptions); err != nil {
+		return nil, err
 	}
 
 	return c, nil
 }
 
+func (c *client) configureAdditionalAccountClient(accounts []*awsv1.AWSAccount, ds *awsv1.ScalingLimits, awsHTTPClient *http.Client, awsClientOptions []func(*config.LoadOptions) error) error {
```
Comment (danielhochman) [inline backend/service/aws/aws.go:109]: nit use `LoadOptionsFunc` for awsClientOptions param

**Review by danielhochman — APPROVED** [review]
lgtm, one nit

File: backend/service/aws/aws.go (hunk 5 above)
Comment (mikecutalo) [reply backend/service/aws/aws.go:81]: Ill try again but the compiler was not having it, thats what I originally lead with.

Comment (mikecutalo) [reply backend/service/aws/aws.go:81]:
```diff
➜ clutch git:(aws-multi-account-gen-clients) ✗ git --no-pager diff
diff --git a/backend/service/aws/aws.go b/backend/service/aws/aws.go
index 36fc93e9..18f90c5c 100644
--- a/backend/service/aws/aws.go
+++ b/backend/service/aws/aws.go
@@ -79,7 +79,7 @@ func New(cfg *anypb.Any, logger *zap.Logger, scope tally.Scope) (service.Service

 	ds := getScalingLimits(ac)
 	awsHTTPClient := &http.Client{}
-	awsClientCommonOptions := []func(*config.LoadOptions) error{
+	awsClientCommonOptions := []config.LoadOptionsFunc{
 		config.WithHTTPClient(awsHTTPClient),
 		config.WithRetryer(func() aws.Retryer {
 			customRetryer := retry.NewStandard(func(so *retry.StandardOptions) {
@@ -107,7 +107,7 @@ func New(cfg *anypb.Any, logger *zap.Logger, scope tally.Scope) (service.Service
 	return c, nil
 }

-func (c *client) configureAdditionalAccountClient(accounts []*awsv1.AWSAccount, ds *awsv1.ScalingLimits, awsHTTPClient *http.Client, awsClientOptions []func(*config.LoadOptions) error) error {
+func (c *client) configureAdditionalAccountClient(accounts []*awsv1.AWSAccount, ds *awsv1.ScalingLimits, awsHTTPClient *http.Client, awsClientOptions []config.LoadOptionsFunc) error {
 	for _, account := range accounts {
 		accountRoleARN := fmt.Sprintf("arn:aws:iam::%s:role/%s", account.AccountNumber, account.IamRole)
 		// For doing STS calls it does not matter which region client we are using, as they are not bounded by region
```

```
building...
# github.com/lyft/clutch/backend/service/aws
service/aws/aws.go:94:10: cannot use append(awsClientCommonOptions, "github.com/aws/aws-sdk-go-v2/config".WithRegion(region)) (type []"github.com/aws/aws-sdk-go-v2/config".LoadOptionsFunc) as type []func(*"github.com/aws/aws-sdk-go-v2/config".LoadOptions) error in argument to "github.com/aws/aws-sdk-go-v2/config".LoadDefaultConfig
service/aws/aws.go:121:11: cannot use append(awsClientOptions, "github.com/aws/aws-sdk-go-v2/config".WithRegion(region)) (type []"github.com/aws/aws-sdk-go-v2/config".LoadOptionsFunc) as type []func(*"github.com/aws/aws-sdk-go-v2/config".LoadOptions) error in argument to "github.com/aws/aws-sdk-go-v2/config".LoadDefaultConfig
failed to build, error: exit status 2
```

File: backend/service/aws/aws.go [hunk 7]
```diff
@@ -63,48 +78,90 @@ func New(cfg *anypb.Any, logger *zap.Logger, scope tally.Scope) (service.Service
...
+			},
+		},
+
+		s3:          s3.NewFromConfig(regionCfg),
+		kinesis:     kinesis.NewFromConfig(regionCfg),
+		ec2:         ec2.NewFromConfig(regionCfg),
+		autoscaling: autoscaling.NewFromConfig(regionCfg),
+		dynamodb:    dynamodb.NewFromConfig(regionCfg),
+		sts:         sts.NewFromConfig(regionCfg),
+	}
```
Comment (scarlettperry) [inline backend/service/aws/aws.go:162]: the former regionalClient struct also had the `iam` field. just checking, do you not need it anymore?

Comment (mikecutalo) [reply backend/service/aws/aws.go:162]: ooo yeah good catch @scarlettperry! Thank you.

Comment (mikecutalo) [reply backend/service/aws/aws.go:162]: It was just added in today https://github.com/lyft/clutch/pull/1863, probably missed it during my rebase.

**Review by scarlettperry — APPROVED** [review]

**Pull request merged**